src/hypofuzz/dashboard.py

```diff
@@ -49,89 +50,242 @@
-        # because it's not progress being made.
-        self.status_counts += counts_diff
-        self.elapsed_time += elapsed_diff
-        status_counts[report.worker.uuid] += counts_diff
-        elapsed_time[report.worker.uuid] += elapsed_diff
-        if report.phase is not Phase.REPLAY:
-            self.reports.append(report)
+
+        # this is 2x the memory in exchange for supporting the by-worker access
+        # pattern. We only access index [-1] though - could we store
+        # self.last_reports_by_worker instead?
+        self.reports_by_worker.setdefault(report.worker.uuid, []).append(linear_report)
+        # Phase.REPLAY does not count towards:
+        #   * status_counts
+        #   * elapsed_time
+        #   * reports
+        #   * branches
+        #   * phase (should we change this? would cause status flipflop with multiple workers)
+        # nor is it displayed on dashboard graphs.
+        # This is fine for display purposes, since these statistics are intended
+        # to convey the time spent searching for bugs. But we should be careful
+        # when measuring cost to compute a separate "overhead" statistic which
+        # takes every input and elapsed_time into account regardless of phase.
+        if linear_report.phase is not Phase.REPLAY:
+            self.linear_reports.append(linear_report)
 
         self._check_invariants()
 
+    @property
     def phase(self) -> Optional[Phase]:
-        if not self.reports:
-            return None
-        return self.reports[-1].phase
+        return self.linear_reports[-1].phase if self.linear_reports else None
+
+    @property
+    def branches(self) -> int:
+        return self.linear_reports[-1].branches if self.linear_reports else 0
+
+    def ninputs(self, since: Optional[float] = None) -> int:
+        return sum(self.linear_status_counts(since=since)[-1].values())
+
+    def _cumsum(
+        self,
+        *,
+        cache: LRUCache[float, tuple[int, list[T]]],
+        attr: str,
+        since: Optional[float],
+        initial: T,
+    ) -> list[T]:
+        if since is None:
+            since = -math.inf
+
+        if since in cache:
+            (start_idx, cumsum) = cache[since]
+            if len(cumsum) < len(self.linear_reports[start_idx:]):
+                # extend cumsum with any new reports
+                running = cumsum[-1] if cumsum else initial
+                for report in self.linear_reports[start_idx + len(cumsum) :]:
+                    value = getattr(report, attr)
+                    assert value >= initial
+                    running += value
+                    cumsum.append(running)
+                cache[since] = (start_idx, cumsum)
+
+            return cumsum
+
+        cumsum = []
+        start_idx = bisect.bisect_right(
+            [report.timestamp_monotonic for report in self.linear_reports], since  # type: ignore
+        )
```

Comment:
Constructing a temporary list like this is O(n), can we use the `.bisect_right()` method on the `self.linear_reports` SortedKeyList instead?